README.rst

```diff
@@ -74,11 +68,167 @@ To install in a virtual environment in your current project:
+
+The server will respond with:
+* Command: Single byte. Always ``0x11``.
+* Status: Single byte.
+* 2 Bytes reserved for padding.
+* Chunk offset: 32-bit number encoding the offset into the file of this chunk.
+* Total length: 32-bit number encoding the total file length.
+* Chunk length: 32-bit number encoding the length of the read data up to the chunk size provided in the header.
+* Chunk-length contents of the file starting from the current position.
+
+If the chunk length is smaller than the total length, then the client will request more data by sending:
+* Command: Single byte. Always ``0x12``.
+* Status: Single byte. Always OK for now.
+* 2 Bytes reserved for padding.
+* Chunk offset: 32-bit number encoding the offset into the file to start the next chunk.
+* Chunk size: 32-bit number encoding the number of bytes to read. May be different than the original size. Does not need to be limited by the total size.
+
+The transaction is complete after the server has replied with all data. (No acknowledgement needed from the client.)
+
+``0x20`` - Write a file
++++++++++++++++++++++++
+
+Writes the content to the given full path. If the file exists, it will be overwritten. Content may be written as received so an interrupted transfer may lead to a truncated file.
+
+Offset larger than the existing file size will introduce zeros into the gap.
+
+The header is four fixed entries and a variable length path:
+
+* Command: Single byte. Always ``0x20``.
+* 1 Byte reserved for padding.
+* Path length: 16-bit number encoding the encoded length of the path string.
+* Offset: 32-bit number encoding the starting offset to write.
+* Total size: 32-bit number encoding the total length of the file contents.
+* Path: UTF-8 encoded string that is *not* null terminated. (We send the length instead.)
+
+The server will repeatedly respond until the total length has been transferred with:
+* Command: Single byte. Always ``0x21``.
+* Status: Single byte. ``0x01`` if OK. ``0x02`` if any parent directory is missing or a file.
+* 2 Bytes reserved for padding.
+* Offset: 32-bit number encoding the starting offset to write. (Should match the offset from the previous 0x20 or 0x22 message)
+* Free space: 32-bit number encoding the amount of data the client can send.
+
+The client will repeatedly respond until the total length has been transferred with:
+* Command: Single byte. Always ``0x22``.
+* Status: Single byte. Always ``0x01`` for OK.
+* 2 Bytes reserved for padding.
+* Offset: 32-bit number encoding the offset to write.
+* Data size: 32-bit number encoding the amount of data the client is sending.
+* Data
+
+The transaction is complete after the server has received all data and replied with a status with 0 free space and offset set to the content length.
+
+
+``0x30`` - Delete a file or directory
++++++++++++++++++++++++++++++++++++++
+
+Deletes the file or directory at the given full path. Directories must be empty to be deleted.
```

Comment:
Maybe we should have an additional/optional parameter to remove folder (recursively). It is rather painful for client to remove 3 level depth of directory (lots of list dir and foreach to remove file). While it could be done very effectively within the native filesytem on server (most likely little filesystem or existing fatfs). I actually have struggled with littlefilesystem and ended up modifying the LFS code to allow this. It is still open though https://github.com/littlefs-project/littlefs/issues/43